src/magic_folder/snapshot.py

```diff
@@ -0,0 +1,172 @@
+# Copyright 2020 Least Authority TFA GmbH
+# See COPYING for details.
+
+"""
+Functions and types that implement snapshots
+"""
+from __future__ import print_function
+
+import time
+import json
+
+from itertools import count
+
+from twisted.internet.defer import (
+    inlineCallbacks,
+    returnValue,
+)
+
+from .common import (
+    get_node_url,
+)
+
+from twisted.web.client import (
+    Agent,
+    readBody,
+)
+
+from twisted.web.http import (
+    OK,
+    CREATED,
+)
+
+from treq.client import (
+    HTTPClient,
+)
+
+from hyperlink import (
+    DecodedURL,
+)
+
+from .common import (
+    bad_response,
+)
+
+
+# version of the snapshot scheme
+SNAPSHOT_VERSION = 1
+
+@inlineCallbacks
+def tahoe_put_immutable(nodeurl, filepath, treq):
```

Comment:
For a public API I think it would be better to pass a file-like object (or maybe our own interface?) instead of a file-path. This falls into the general idea of "pass the thing you actually need, not information to get the thing".

Passing a file-like would make testing easier and also not tie the API to "a filesystem" (e.g. what if I want to upload a JSON document that I just serialized from a `dict` in-memory).

Reply:
Yes, passing "file like" things instead of file paths sounds very good. .